ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+            next_state = next_state()
+
+    def _state_detect_boundary(self):
+        next_state = None
+        match = self._boundary_rgx.search(self._buffer)
+        if match:
+            # This is here as a paranoia check; we should only be in this state when the buffer
+            # immediately starts with the boundary line.
+            span = match.span()
+            data_before = self._buffer[:span[0]]
+            if data_before:
+                raise RuntimeError(
+                    'Unexpected data before boundary: {!r}'.format(data_before))
+
+            is_terminating_boundary = match.group(1)
+            if is_terminating_boundary:
+                # Permanently stop the FSM, ignoring any remaining data
+                next_state = self._state_stop
+                self._update_initial_state(next_state)
+            else:
+                # Consume the boundary and get ready for parsing the headers.
+                self._buffer = self._buffer[span[1]:]
+                next_state = self._state_parse_header
+        else:
+            # Couldn't find a complete boundary in the buffer; stop for now
+            self._update_initial_state(self._state_detect_boundary)
+            next_state = self._state_stop
+        return next_state
+
+    def _state_parse_header(self):
+        next_state = None
+        try:
+            headers, content = self._buffer.split(b'\r\n\r\n', 1)
+        except ValueError:
+            # Not enough data; stop for now
+            self._update_initial_state(self._state_parse_header)
+            next_state = self._state_stop
+        else:
+            # Rely on the stdlib for parsing the headers, since this can be done purely
+            # in memory without any realistic worry of memory concerns.
+            parser = email.parser.BytesFeedParser()
+            parser.feed(headers)
+            message = parser.close()
+
+            content_disposition = message.get_content_disposition()
+            if content_disposition != 'form-data':
+                raise RuntimeError(
+                    'unexpected content disposition: {}'.format(content_disposition))
+
+            name = message.get_param('name', header='content-disposition')
+            if name == 'response':
+                next_state = self._state_parse_response_part
+            elif name == 'files':
+                filename = message.get_filename()
+                if filename is None:
+                    raise NotImplementedError('multipart "files" part missing filename')
+                self._prepare_tempfile(filename)
+                next_state = self._state_parse_files_part
+            else:
+                raise RuntimeError(
+                    'unexpected name in content-disposition header: {}'.format(name))
+
+            # Consume the header; now we're ready to parse the part body
+            # (either "request" or "files")
+            self._buffer = content
+        return next_state
```

Comment:
I didn't get past this point for this review. I definitely like the design of the state machine, I feel like we are just really close to having a small buffer file object rather than buffering everything to temp.

Reply:
The FSM was dumped in favor of a stateful loop implementation.